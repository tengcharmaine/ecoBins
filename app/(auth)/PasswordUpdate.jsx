import React, { useState } from "react";
import { View, Text, TextInput, Button } from "react-native";
import { supabase } from "../../lib/supabase";
import { useRoute } from '@react-navigation/native';


export default function PasswordUpdate() {
  const route = useRoute();

  const { userId } = route.params;

  const [password, setPassword] = useState("");
  const [loading, setLoading] = useState(false);
  const [errMsg, setErrMsg] = useState("");

  const handlePasswordUpdate = async () => {
    setErrMsg("");

    if (password === "") {
      setErrMsg("Password cannot be empty");
      return;
    }

    //const { data: { user }, error: userError } = await supabase.auth.getUser()
    setLoading(true);
    const { error } = await supabase.auth.updateUser(userId, {
      password: password,
    });
    setLoading(false);

    if (error) {
      setErrMsg(error.message);
      return;
    }
  };

  return (
    <View>
      <Text>Enter your new password:</Text>
      <TextInput
        placeholder="New Password"
        value={password}
        onChangeText={setPassword}
        secureTextEntry
      />
      <Button onPress={handlePasswordUpdate} title="Update Password" />
      {errMsg !== "" && <Text>{errMsg}</Text>}
      {loading && <ActivityIndicator />}
    </View>
  );
}
