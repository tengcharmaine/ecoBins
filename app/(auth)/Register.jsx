import { useState } from "react";
import { supabase } from "../../lib/supabase";
import { StyleSheet, View, Alert } from "react-native";
import { Text, TextInput, ActivityIndicator, Button } from 'react-native-paper';
import { useNavigation } from '@react-navigation/native';
import { KeyboardAwareScrollView } from "react-native-keyboard-aware-scroll-view";


export default function Register() {
    const navigation = useNavigation();
    const [email, setEmail] = useState('');
    const [password, setPassword] = useState('');
    const [loading, setLoading] = useState(false);
    const [errMsg, setErrMsg] = useState('');
    const [emailErrMsg, setEmailErrMsg] = useState('');
    const [passwordErrMsg, setPasswordErrMsg] = useState('');
    const [showEmailCheck, setShowEmailCheck] = useState(false);

    

    const handleSubmit = async () => {
        setErrMsg('');
        setEmailErrMsg('');
        if (email == '') {
            setEmailErrMsg("Email cannot be empty")
            return;
        }
        setPasswordErrMsg('');
        if (password == '') {
            setPasswordErrMsg("Password cannot be empty")
            return;
        }
    
        setLoading(true);
        const { error } = await supabase.auth.signUp({ email, password });
        setEmail(email);
        setLoading(false);
        if (error) {
            setErrMsg(error.message);
            return;
        }
        console.log(email)

        setShowEmailCheck(true); // Show the "Check your email" message
   
        navigation.navigate('Login');
    };



    const showVerificationAlert = () => {
                Alert.alert(
                    'Email Verification',
                    'Please check your email for the verification link.',
                    [
                        { text: 'OK' },
                    ]
                );
            };
    const styles = StyleSheet.create({
        container: {
            flex: 1, 
            justifyContent: 'center', 
        },
        innerContainer: {
            justifyContent: 'center',
            alignItems: 'center',
            marginTop: '10%', // Adjust the margin as needed
            marginBottom: 30, 
        },
        input: {
            borderColor: "black",
            borderWidth: 1,
            backgroundColor: "white",
            width: '75%',
            borderRadius: 5
        },
        button: {
            borderColor: "black",
            alignItems: 'center',
            backgroundColor: "#c7dede",
            width: '25%',
            marginTop: 20,
            marginBottom: 10,
            borderRadius: 10,
            
        },
        text1: {
            color: "black",
            marginTop: 20,
            textAlign: 'left',
            marginRight: 255,
            marginBottom: 5,
        },

        text2: {
            color: "black",
            marginTop: 20,
            textAlign: 'left',
            marginRight: 230,
            marginBottom: 5,
        },

        title: {
            color: "black",
            fontSize: 20,
            marginBottom: 20, 
            fontWeight: "bold",
        },
        error: {
            color: "red",
            marginTop: 4,
            marginBottom: 5,
        },
        
    });

    return (
    <KeyboardAwareScrollView
      contentContainerStyle={styles.container}
      resetScrollToCoords={{ x: 0, y: 0 }}
      scrollEnabled={true}>
        <View style={styles.innerContainer}>
            <Text style= {styles.title}> Registration </Text>
            <Text style= {styles.text1}>Email</Text>
            <TextInput
                autoCapitalize='none'
                placeholder="Email"
                placeholderTextColor={"#dfd8dc"}
                style={styles.input}
                textContentType='emailAddress'
                value={email}
                onChangeText={setEmail} />
            {emailErrMsg !== "" && <Text style= {styles.error}>{emailErrMsg}</Text>}

            <Text style= {styles.text2}>Password</Text>
            <TextInput
                secureTextEntry
                placeholder="Password"
                placeholderTextColor={"#dfd8dc"}
                style={styles.input}
                autoCapitalize='none'
                textContentType='password'
                value={password}
                onChangeText={setPassword} />
            {passwordErrMsg !== "" && <Text style= {styles.error}>{passwordErrMsg}</Text>}

            <Button style = {styles.button} onPress={handleSubmit}>
                <Text style={styles.text1}> Enter </Text>
            </Button>          
            {errMsg !== "" && <Text style= {styles.error}>{errMsg}</Text>}
            {loading && <ActivityIndicator />}
            {showEmailCheck && showVerificationAlert()}
        </View>
    </KeyboardAwareScrollView>
    );
}