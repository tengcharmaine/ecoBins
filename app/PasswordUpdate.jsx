import React, { useState, useEffect } from "react";
import { View, StyleSheet, TouchableOpacity, Image  } from "react-native";
import { Text, TextInput, ActivityIndicator, Button, IconButton} from 'react-native-paper';
import { supabase } from "../lib/supabase";
import { useRoute, useNavigation } from '@react-navigation/native';
import { KeyboardAwareScrollView } from "react-native-keyboard-aware-scroll-view";
import { Ionicons } from '@expo/vector-icons';
import { useFonts } from 'expo-font';
import {useRouter} from 'expo-router';



import * as ExpoLinking from 'expo-linking'; // Import ExpoLinking from the 'expo-linking' package
import { useURL } from 'expo-linking'; // Import the useURL hook from 'expo-linking'
import * as Linking from 'expo-linking';


export default function PasswordUpdate() {
  const route = useRoute();
  const navigation = useNavigation();
  const router = useRouter();


  const [loaded] = useFonts({
    Poppins: require('../assets/fonts/Poppins-Regular.ttf'),
    PoppinsBold: require('../assets/fonts/Poppins-Bold.ttf'),
    PoppinsSemiBold: require('../assets/fonts/Poppins-SemiBold.ttf'),
    PoppinsBlack: require('../assets/fonts/Poppins-Black.ttf'),
  });

  //const { userId, token } = route.params;
  //console.log(userId);
  console.log(route.params);
  const [password, setPassword] = useState("");
  const [loading, setLoading] = useState(false);
  const [errMsg, setErrMsg] = useState("");
  const [passwordVisible, setPasswordVisible] = useState(false);

  const handleReturn = () => {
    navigation.goBack();
  };

  const togglePasswordVisibility = () => {
    setPasswordVisible(!passwordVisible);
  }

  const handlePasswordUpdate = async () => {
    setErrMsg("");

    if (password === "") {
      setErrMsg("Password cannot be empty");
      return;
    }

    //const { data: { user }, error: userError } = await supabase.auth.getUser()
    setLoading(true);
    const { error } = await supabase.auth.updateUser({
      password: password,
    });
    setLoading(false);

    if (error) {
      setErrMsg(error.message);
      return;
    }

    navigation.navigate("passwordUpdateSuccess");
  };

  const goBack = () => {
    navigation.navigate('settings');
  }

  const styles = StyleSheet.create({
    container: {
        flex: 1, 
        backgroundColor: "white",

    },
    input: {
      flexDirection: 'row',
      alignItems: 'flex-end',
      marginBottom: 20,
    },
    title: {
      color: "black",
      marginTop: 20,
      justifyContent: 'center',
      fontWeight: 'bold',
      fontSize: 34,
      textAlign: 'left',
      marginRight: 40,
      marginBottom: 15,
      fontFamily: 'PoppinsBold',
  },
    button: {
      borderColor: "black",
      alignItems: 'center',
      justifyContent: 'center',
      backgroundColor: "#c7dede",
      width: '85%',
      height: 60,
      marginTop: 20,
      marginBottom: 10,
      borderRadius: 20,
        
    },
    text1: {
      color: "black",
      fontWeight: 'bold',
      fontSize: 19,
      textAlign: 'center',
      fontFamily: 'PoppinsSemiBold',
    },
  
    text2: {
        color: "black",
        textAlign: "left",
    },
  
    text3: {
        color: "black",
        marginTop: 20,
        textAlign: 'left',
        marginRight: 255,
        marginBottom: 5,
    },
    passwordInput: {
      borderColor: "white",
      borderWidth: 1,
      borderBottomColor: "grey",
      backgroundColor: "white",
      width: '75%',
    },
  
    error: {
        color: "red",
        marginTop: 4,
        marginBottom: 5,
    },
    innerContainer: {
        justifyContent: 'center',
        alignItems: 'center',
        marginTop: '80%', 
        marginBottom: 30, 
    },
    passwordIcon: {
      position: 'absolute',
      right: 10,
      top: '30%',
      transform: [{ translateY: -12 }],
    },
    backButton: {
      position: 'absolute',
      top: 60,
      left: 20,
      zIndex: 1,
  },
    emailIcon: {
      width: 30,
      height: 30,
      resizeMode: 'contain',
      marginRight: 20,
    },
    error: {
      color: "red",
      fontFamily: 'Poppins',
      //marginTop: 10,
  },
  });

  if (!loaded) {
    return (
      <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
        <ActivityIndicator size="large" color="black" />
      </View>
    );
  }

  return (
    <KeyboardAwareScrollView
      contentContainerStyle={styles.container}
      resetScrollToCoords={{ x: 0, y: 0 }}
      scrollEnabled={true}
    >
    <TouchableOpacity onPress={goBack} style={styles.backButton}>
      <Ionicons name="arrow-back" size={24} color="black" />
    </TouchableOpacity>
    <View style= {styles.innerContainer}>
    <Text style= {styles.title}> Reset password </Text>
      {/* <Text style= {styles.text1}>Password</Text> */}
            <View style={styles.input}>
            <Image
            source={require('./../images/lock.png')}
            style={styles.emailIcon}
          />
                <TextInput
                    secureTextEntry={!passwordVisible}
                    placeholder="Password"
                    placeholderTextColor={"#dfd8dc"}
                    style={styles.passwordInput}
                    autoCapitalize='none'
                    textContentType='password'
                    value={password}
                    onChangeText={setPassword}
                />
                <TouchableOpacity
                    style={styles.passwordIcon}
                    onPress={togglePasswordVisibility}
                >
                <IconButton
                    icon={passwordVisible ? "eye-off" : "eye"}
                    color="#000"
                    size={20}
                />
                </TouchableOpacity>
            </View>
            {errMsg !== "" && <Text style = {styles.error}>{errMsg}</Text>}

            <TouchableOpacity style = {styles.button} onPress={handlePasswordUpdate}>
                <Text style={styles.text1}> Update password </Text>
            </TouchableOpacity>
      {loading && <ActivityIndicator />}
    </View>
    </KeyboardAwareScrollView>
  );
}