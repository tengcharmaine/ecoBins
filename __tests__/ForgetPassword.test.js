import React from 'react';
import renderer from 'react-test-renderer';
import Register from '../components/register';
import { act } from 'react-dom/test-utils';
import PasswordUpdate from '../components/passwordupdate';
import ForgetPasswordScreen from '../components/forgetpassword';


describe('Register', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });
  test('renders correctly', () => {
    const tree = renderer.create(<ForgetPasswordScreen />).toJSON();
    
    // Check if the required elements are rendered
    expect(tree).toMatchSnapshot();
  });

  test('displays error messages for empty credentials', () => {
    const component = renderer.create(<Register />);
    const instance = component.getInstance();

    // Simulate form submission without entering any values
    instance.handleSubmit();

    act(() => {
      // Advance timers manually
      jest.advanceTimersByTime(1000); // Advances timers by 1 second
    });

    // Check if the error messages are displayed
    expect(instance.state.emailErrMsg).toBe('Email cannot be empty');
    expect(instance.state.passwordErrMsg).toBe('Password cannot be empty');
  });

  test('sends email with a valid email address', async () => {
    const tree = renderer.create(<ForgetPasswordScreen />);
    const instance = tree.getInstance();
  
    // Set a valid email address
    const validEmail = 'example@example.com';
    instance.setState({ email: validEmail });
  
    // Simulate the "Continue" button click
    await act(async () => {
      await instance.handleSendEmail();
    });
  
    // Check if the loading state is set to false after email sending
    expect(instance.state.loading).toBe(false);
  });
  
  test('displays error message for empty email address', async () => {
    const tree = renderer.create(<ForgetPasswordScreen />);
    const instance = tree.getInstance();
  
    // Set an empty email address
    const emptyEmail = '';
    instance.setState({ email: emptyEmail });
  
    // Simulate the "Continue" button click
    await act(async () => {
      await instance.handleSendEmail();
    });
  
    // Check if the error message for empty email address is displayed
    expect(instance.state.emailErrMsg).toBe('Email cannot be empty');
  });

  test('verifies OTP with a valid OTP', async () => {
    const tree = renderer.create(<ForgetPasswordScreen />);
    const instance = tree.getInstance();
  
    // Set a valid OTP
    const validOTP = '123456';
    instance.setState({ otp: validOTP, isEmailSent: true });
  
    // Simulate the "Verify OTP" button click
    await act(async () => {
      await instance.handleVerifyOTP();
    });
  
    // Check if the loading state is set to false after OTP verification
    expect(instance.state.loading).toBe(false);
  });

  test('displays error message for invalid OTP', async () => {
    const tree = renderer.create(<ForgetPasswordScreen />);
    const instance = tree.getInstance();
  
    // Set an invalid OTP
    const invalidOTP = '987654';
    instance.setState({ email: 'test@gmail.com', otp: invalidOTP, isEmailSent: true });
  
    // Simulate the "Verify OTP" button click
    await act(async () => {
      await instance.handleVerifyOTP();
    });
  
    // Check if the error message for invalid OTP is displayed
    expect(instance.state.error).toBe('Token has expired or is invalid');
  });
  
  
  
});