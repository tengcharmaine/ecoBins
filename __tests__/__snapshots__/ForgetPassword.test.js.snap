// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "rgba(28, 27, 31, 1)",
            "fontFamily": "System",
            "fontWeight": "400",
            "letterSpacing": 0,
          },
          {
            "writingDirection": "ltr",
          },
          {
            "color": "black",
            "fontSize": 34,
            "fontWeight": "bold",
            "justifyContent": "center",
            "marginBottom": 15,
            "marginRight": 50,
            "marginTop": 20,
            "textAlign": "left",
          },
        ]
      }
    >
      Forget Password
    </Text>
    <View
      style={
        {
          "alignItems": "flex-end",
          "flexDirection": "row",
          "marginBottom": 20,
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "rgba(231, 224, 236, 1)",
              "borderTopLeftRadius": 4,
              "borderTopRightRadius": 4,
            },
            {
              "backgroundColor": "white",
              "borderBottomColor": "grey",
              "borderColor": "white",
              "borderWidth": 1,
              "marginTop": 15,
              "width": 260,
            },
          ]
        }
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "rgba(28, 27, 31, 1)",
              "bottom": 0,
              "height": 1,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "transform": [
                {
                  "scaleY": 0.5,
                },
              ],
              "zIndex": 1,
            }
          }
          testID="text-input-underline"
        />
        <View
          style={
            [
              {
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              {
                "minHeight": 56,
              },
            ]
          }
        >
          <TextInput
            autoCapitalize="none"
            editable={true}
            maxFontSizeMultiplier={1.5}
            multiline={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="Email"
            placeholderTextColor="#dfd8dc"
            selectionColor="rgba(103, 80, 164, 1)"
            style={
              [
                {
                  "margin": 0,
                },
                {
                  "paddingLeft": 16,
                  "paddingRight": 16,
                },
                {
                  "height": 56,
                },
                {
                  "paddingBottom": 0,
                  "paddingTop": 0,
                },
                {
                  "color": "rgba(73, 69, 79, 1)",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": undefined,
                  "letterSpacing": 0.15,
                  "lineHeight": undefined,
                  "textAlign": "left",
                  "textAlignVertical": "center",
                },
                false,
                [
                  {},
                ],
                undefined,
              ]
            }
            testID="text-input-flat"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
      </View>
    </View>
    <View
      collapsable={false}
      style={
        {
          "alignSelf": "center",
          "backgroundColor": "#c7dede",
          "borderRadius": 20,
          "height": 60,
          "marginBottom": 10,
          "marginTop": 20,
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
          "width": 320,
        }
      }
      testID="button-container-outer-layer"
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#c7dede",
            "borderColor": "black",
            "borderRadius": 20,
            "borderStyle": "solid",
            "borderWidth": 0,
            "flex": 1,
            "justifyContent": "center",
            "minWidth": 64,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="button-container"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              {
                "borderRadius": 20,
              },
            ]
          }
          testID="button"
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                [
                  {
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "letterSpacing": 0.1,
                    "lineHeight": 20,
                  },
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    false,
                    {
                      "marginHorizontal": 12,
                    },
                    undefined,
                    false,
                    {
                      "color": "rgba(103, 80, 164, 1)",
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    undefined,
                  ],
                ]
              }
              testID="button-text"
            >
              <Text
                style={
                  [
                    {
                      "textAlign": "left",
                    },
                    {
                      "color": "rgba(28, 27, 31, 1)",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "letterSpacing": 0,
                    },
                    {
                      "writingDirection": "ltr",
                    },
                    {
                      "color": "black",
                      "fontSize": 19,
                      "fontWeight": "bold",
                      "textAlign": "center",
                    },
                  ]
                }
              >
                Continue
              </Text>
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
